mesonbuild/mesonlib.py

```diff
@@ -34,6 +34,8 @@
     from .compilers.compilers import CompilerType
     from .interpreterbase import ObjectHolder
 
+    FileOrString = T.Union['File', str]
```

Comment:
Is there a specific reason/benefit why this is inside the `T.TYPE_CHECKING` if while the? If not, would you consider moving this out, so that it doesn't have to be quoted in other parts of the codebase?

IMHO it isn't really intuitive what style is required (especially for new python developers) if `T.TYPE_CHECKING` blocks in other files have to be checked.

Reply:
Ḃecause it's only used by mypy? It seems really strange to me to define or import things outside of an if type_checking block that are only used for type checking.

Reply:
The thing is, we are currently doing this inconsistently and I dislike using strings because of missing syntax highlighting. Also, you could argue that it is (kind of) used in the type hints since they are part of the code. Python will complain if you have a typo in an annotation without strings for instance.

Reply:
Until 3.7 when they realized it was a terrible idea for the interpreter to actually interpret the type hints (because it has memory and startup time considerations). in the 3.7 with `__future__.annotations` the interpreter treats them as strings always. Eventually (like all __future__ stuff) that will become the default. I wish we could drop 3.6 support already and just use the 3.7+ syntax :)